File: MealTracker/MealTrackerTableViewController.m
//
//  MealTrackerTableViewController.m
//  MealTracker
//
//  Created by Will Hindenburg on 7/21/12.
//  Copyright (c) 2012 __MyCompanyName__. All rights reserved.
//

#import <BaseClassesSDK/User.h>

#import "MealTrackerTableViewController.h"
#import "MealEntryViewController.h"
#import "SaveMealService.h"
#import "RetrieveMealService.h"
#import "MealEaten.h"
#import "MealEatenService.h"
#import "Meal.h"
#import "DeleteMealService.h"

@interface MealTrackerTableViewController () <MealTextEntryDelegate>
@property (nonatomic, strong) NSArray *dataSource;
@property (strong, nonatomic) User *user;
@end


@implementation MealTrackerTableViewController

- (void)viewController:(id)sender didFinishWithMeal:(Meal *)meal {
    [[SaveMealService new] saveMeal:meal withSuccessBlock:^(Meal *meal) {
        [self.tableView reloadData];
    } andError:^(NSError *error) {
        [super showError:error withRetryBlock:^{
            [self viewController:sender didFinishWithMeal:meal];
        }];
    }];
}

- (void)viewController:(id)sender didFinishEditingMeal:(Meal *)meal {
    SaveMealService *saveService = [SaveMealService new];
    [saveService updateMeal:meal withSuccessBlock:^{
        
    } andError:^(NSError *error) {
        [super showError:error withRetryBlock:^{
            [self viewController:sender didFinishEditingMeal:meal];
        }];
    }];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    UINavigationController *navController = (UINavigationController *)segue.destinationViewController;
    [(MealEntryViewController *)navController.topViewController setTextEntryDelegate:self];
    
    if ([sender isKindOfClass:[UITableViewCell class]])
    {
        UITableViewCell *cell = sender;
        NSIndexPath *indexPath = [self.tableView indexPathForCell:cell];

        [(MealEntryViewController *)navController.topViewController setMealData:[self.dataSource objectAtIndex:indexPath.row]];
    }
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [self retrieveMealsFromService];
    
}

- (void)retrieveMealsFromService {
    RetrieveMealService *mealService = [RetrieveMealService new];
    [super showActivityIndicatorAnimated:YES];
    [mealService retrieveMealsWithSuccessBlock:^(NSArray *meals) {
        self.dataSource = meals;
        [self.tableView reloadData];
        [super hideActivityIndicatorAnimated:YES];
    } andError:^(NSError *error) {
        [super hideActivityIndicatorAnimated:YES];
        [super showError:error withRetryBlock:^{
            [self retrieveMealsFromService];
        }];
    }];
}

- (void)viewDidLoad
{
    [super viewDidLoad];

    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
 
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    self.navigationItem.leftBarButtonItem = self.editButtonItem;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    self.user = [User persistentUserObject];
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"Meal Cell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
        
    }
    UIButton *button = [[UIButton alloc] initWithFrame:CGRectMake(0, 0, 35, 35)];
    [button setBackgroundImage:[UIImage imageNamed:@"plusButton"] forState:UIControlStateNormal];
    [button addTarget:self action:@selector(disclosureButtonPressed:) forControlEvents:UIControlEventTouchUpInside];
    cell.accessoryView = button;
    // ask NSFetchedResultsController for the NSMO at the row in question
    Meal *meal = [self.dataSource objectAtIndex:indexPath.row];
    // Then configure the cell using it ...
    cell.textLabel.text = meal.name;
    cell.detailTextLabel.text = [NSString stringWithFormat:@"%@ Weight Watchers Points", [meal.weightWatchersPlusPoints stringValue]];
    
    return cell;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        Meal *meal = [self.dataSource objectAtIndex:indexPath.row];
        [super showActivityIndicatorAnimated:YES];
        [[DeleteMealService new] removeMeal:meal withSuccessBlock:^{
            NSMutableArray *mutableCapsuleArray = [self.dataSource mutableCopy];
            [mutableCapsuleArray removeObjectAtIndex:indexPath.row];
            self.dataSource = [mutableCapsuleArray copy];
            [self.tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
            [super hideActivityIndicatorAnimated:YES];
        } andError:^(NSError *error) {
            [super hideActivityIndicatorAnimated:YES];
            [super showError:error withRetryBlock:^{
                [self tableView:tableView commitEditingStyle:editingStyle forRowAtIndexPath:indexPath];
            }];
        }];
    }
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.dataSource.count;
}

#pragma mark - Table view delegate

- (void)disclosureButtonPressed:(id)sender
{
    UITableViewCell *cell = (UITableViewCell *)[sender superview];
    NSIndexPath *indexPath = [self.tableView indexPathForCell:cell];
    
    UIImageView *imageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"plusButton"]];
    
    CGPoint p = [self.tableView convertPoint:cell.accessoryView.frame.origin fromView:cell];
    imageView.frame = CGRectMake(p.x, p.y, 35, 35);
    [self.view addSubview:imageView];
    
    
    [UIView animateWithDuration:1.0 delay:0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
        imageView.frame = CGRectMake(self.view.frame.size.width - ((self.view.frame.size.width - cell.accessoryView.frame.origin.x) + 50.0), (self.tableView.bounds.size.height + self.tableView.contentOffset.y), imageView.frame.size.width, imageView.frame.size.height);
        imageView.transform = CGAffineTransformScale(CGAffineTransformIdentity, 0.99, 0.99);
    } completion:^(BOOL finished) {
        imageView.transform = CGAffineTransformIdentity;
        [imageView removeFromSuperview];
    }];

    
    MealEaten *mealEaten = [MealEaten new];
    mealEaten.meal = [self.dataSource objectAtIndex:indexPath.row];
    mealEaten.user = self.user;
    mealEaten.dateEaten = [NSDate date];
    [self saveMealEaten:mealEaten];
    
}

- (void)saveMealEaten:(MealEaten *)mealEaten {
    [super showActivityIndicatorAnimated:YES];
    [[MealEatenService new] saveMealEaten:mealEaten withSuccessBlock:^(MealEaten *mealEaten) {
        [super hideActivityIndicatorAnimated:YES];
    } andError:^(NSError *error) {
        [super hideActivityIndicatorAnimated:YES];
        [super showError:error withRetryBlock:^{
            [self saveMealEaten:mealEaten];
        }];
    }];
}

@end

File: MealTracker/MealTrackerTableViewController.swift
//
//  MealTrackerTableViewController.swift
//  MealTracker
//
//  Created by William Hindenburg on 8/29/15.
//
//

import UIKit
import BaseClassesSwift

class MealTrackerTableViewController: MealBaseTableViewController, MealTextEntryDelegate {
    var dataSource:Array <Meal>?
    var user:User?
    
    func viewController(_ sender: UIViewController, didFinishWithMeal: Meal) {
        super.showActivityIndicator(animated: true)
        SaveMealServiceSwift().saveMeal(didFinishWithMeal, successBlock: { (meal) -> Void in
            self.tableView.reloadData()
            super.hideActivityIndicator(animated: true)
        }) { (error) -> Void in
            super.hideActivityIndicator(animated: true)
            super.showError(error, withRetry: { () -> Void in
                self.viewController(sender, didFinishWithMeal: didFinishWithMeal)
            })
        }
    }
    
    func viewController(_ sender: UIViewController, didFinishEditingMeal: Meal) {
        super.showActivityIndicator(animated: true)
        SaveMealServiceSwift().updateMeal(didFinishEditingMeal, successBlock: { () -> Void in
            super.hideActivityIndicator(animated: true)
        }) { (error) -> Void in
            super.hideActivityIndicator(animated: true)
            super.showError(error, withRetry: { () -> Void in
                self.viewController(sender, didFinishEditingMeal: didFinishEditingMeal)
            })
        }
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        let navController = segue.destination as! UINavigationController
        let mealEntryController = navController.topViewController as! MealEntryTableViewController
        mealEntryController.textEntryDelegate = self
        
        if let cell = sender as? UITableViewCell {
            let indexPath = self.tableView.indexPath(for: cell)
            if indexPath != nil {
                mealEntryController.mealData = self.dataSource?[indexPath!.row]
            }
            
            
        }
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        self.retrieveMealsFromService()
    }
    
    func retrieveMealsFromService() {
        super.showActivityIndicator(animated: true)
        RetrieveMealService().retrieveMeals({ (meals) -> Void in
            self.dataSource = meals
            self.tableView.reloadData()
            super.hideActivityIndicator(animated: true)
        }) { (error) -> Void in
            super.hideActivityIndicator(animated: true)
            super.showError(error, withRetry: { () -> Void in
                self.retrieveMealsFromService()
            })
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.navigationItem.leftBarButtonItem = self.editButtonItem
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(true)
        self.user = User.persistentUserObject()
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cellIdentifier = "Meal Cell"
        
        var cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier)
        
        if cell == nil {
            cell = UITableViewCell(style: UITableViewCellStyle.default, reuseIdentifier: cellIdentifier)
        }
        let mealCell = cell!
        
        let button = UIButton(frame: CGRect(x: 0, y: 0, width: 35, height: 35))
        button.setBackgroundImage(UIImage(named: "plusButton"), for: UIControlState())
        button.addTarget(self, action: #selector(MealTrackerTableViewController.disclosureButtonPressed(_:)), for: UIControlEvents.touchUpInside)
        mealCell.accessoryView = button
        
        let meal = self.dataSource?[indexPath.row]
        mealCell.textLabel?.text = meal?.name
        mealCell.detailTextLabel?.text = (meal?.weightWatchersPlusPoints.stringValue)! + " Weight Watchers Points"
        
        return mealCell
        
    }
    
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == UITableViewCellEditingStyle.delete {
            let meal = self.dataSource?[indexPath.row]
            super.showActivityIndicator(animated: true)
            DeleteMealServiceSwift().removeMeal(meal!, successBlock: { () -> Void in
                self.dataSource?.remove(at: indexPath.row)
                self.tableView.deleteRows(at: [indexPath], with: UITableViewRowAnimation.automatic)
                super.hideActivityIndicator(animated: true)
            }, errorBlock: { (error) -> Void in
                super.hideActivityIndicator(animated: true)
                super.showError(error, withRetry: { () -> Void in
                    self.tableView(tableView, commit: editingStyle, forRowAt: indexPath)
                })
            })
        }
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.dataSource?.count ?? 0
    }
    
    func disclosureButtonPressed(_ sender:UIButton) {
        if let cell = sender.superview as? UITableViewCell {
            let imageView = UIImageView(image: UIImage(named: "plusButton"))
            let p = self.tableView.convert((cell.accessoryView?.frame.origin)!, from: cell)
            imageView.frame = CGRect(x: p.x, y: p.y, width: 35, height: 35)
            self.view.addSubview(imageView)
            
            UIView.animate(withDuration: 1.0, delay: 0, options: UIViewAnimationOptions(), animations: { () -> Void in
                imageView.frame = CGRect(x: self.view.frame.size.width - ((self.view.frame.size.width - cell.accessoryView!.frame.origin.x) + 50.0), y: (self.tableView.bounds.size.height + self.tableView.contentOffset.y), width: imageView.frame.size.width, height: imageView.frame.size.height);
                imageView.transform = CGAffineTransform.identity.scaledBy(x: 0.99, y: 0.99);
            }, completion: { (finished) -> Void in
                imageView.transform = CGAffineTransform.identity;
                imageView.removeFromSuperview()
            })
            
            if let indexPath = self.tableView.indexPath(for: cell) {
                let mealEaten = MealEatenPost()
                mealEaten.meal = (self.dataSource?[indexPath.row])!
                mealEaten.dateEaten = Date()
                self.saveMealEaten(mealEaten)
            }            
        }
    }
    
    func saveMealEaten(_ mealEaten:MealEatenPost) {
        super.showActivityIndicator(animated: true)
        MealEatenServiceSwift().saveMealEaten(mealEaten, successBlock: { (meal) -> Void in
            super.hideActivityIndicator(animated: true)
        }) { (error) -> Void in
            super.hideActivityIndicator(animated: true)
            super.showError(error, withRetry: { () -> Void in
                self.saveMealEaten(mealEaten)
            })
        }
    }

}
